test/hotspot/jtreg/testlibrary_tests/verify/tests/TestVerify.java

```diff
@@ -400,20 +462,233 @@ public static void testRecursive() {
+        F2 f3 = new F2(2, 1);
+        F2 f4 = new F2(1, 2);
+
+        Verify.checkEQ(f1, f1, false, true);
+        Verify.checkEQ(f1, f2, false, true);
+        Verify.checkEQ(f2, f1, false, true);
+        checkNE(f1, f3, false, true);
+        checkNE(f1, f4, false, true);
+        checkNE(f3, f1, false, true);
+        checkNE(f4, f1, false, true);
+        checkNE(f3, f4, false, true);
+        checkNE(f4, f3, false, true);
+
+        G g1 = new G(1.0f, 1.0f);
+        G g2 = new G(1.0f, 1.0f);
+        G g3 = new G(Float.NaN, Float.NaN);
+        G g4 = new G(Float.NaN, Float.NaN);
+
+        Verify.checkEQ(g1, g1, false, true);
+        Verify.checkEQ(g2, g1, false, true);
+        Verify.checkEQ(g1, g2, false, true);
+        Verify.checkEQ(g3, g3, false, true);
+        Verify.checkEQ(g3, g4, false, true);
+        Verify.checkEQ(g4, g3, false, true);
+        checkNE(g1, g3, false, true);
+        checkNE(g3, g1, false, true);
+
+        // Nested class with primitive types, where the boxed types may not be cached,
+        // and so they would create different boxed objects.
+        Verify.checkEQ(new H2(), new H2(), false, true);
+
+        // Records.
+        R1 r11 = new R1();
+        R1 r12 = new R1();
+        R2 r21 = new R2();
+        R3 r31 = new R3(1, 1);
+        R3 r32 = new R3(1, 1);
+        R3 r33 = new R3(1, 2);
+        R3 r34 = new R3(2, 1);
+
+        Verify.checkEQ(r11, r11, false, true);
+        Verify.checkEQ(r11, r12, false, true);
+        Verify.checkEQ(r12, r11, false, true);
+        checkNE(r11, r21, false, true);
+        Verify.checkEQ(r31, r31, false, true);
+        Verify.checkEQ(r31, r32, false, true);
+        Verify.checkEQ(r32, r31, false, true);
+        checkNE(r31, r33, false, true);
+        checkNE(r33, r31, false, true);
+        checkNE(r31, r34, false, true);
+        checkNE(r34, r31, false, true);
+        checkNE(r33, r34, false, true);
+
+        R4 r41 = new R4(null, null);
+        R4 r42 = new R4(null, null);
+        R4 r43 = new R4(r41, null);
+        R4 r44 = new R4(r42, null);
+        R4 r45 = new R4(r43, r41);
+        R4 r46 = new R4(r44, r42);
+        R4 r47 = new R4(r44, r41);
+
+        Verify.checkEQ(r45, r46, false, true);
+        Verify.checkEQ(r46, r45, false, true);
+        checkNE(r45, r47, false, true);
+        checkNE(r47, r45, false, true);
+        checkNE(r46, r47, false, true);
+        checkNE(r47, r46, false, true);
+    }
+
+    public static void checkNE(Object a, Object b, boolean isFloatCheckWithRawBits, boolean isCheckWithArbitraryClasses) {
          try {
-            Verify.checkEQ(a, b);
-            throw new RuntimeException("Should have thrown");
+            Verify.checkEQ(a, b, isFloatCheckWithRawBits, isCheckWithArbitraryClasses);
+            throw new RuntimeException("Should have thrown: " + a + " vs " + b);
         } catch (VerifyException e) {}
     }
+
+    public static void checkNE(Object a, Object b) {
```

Comment:
If `checkEQ` lives in Verify, wouldn't it make sense to also have `checkNE` there? Seems like the natural place making the API symmetric.

Reply:
To be honest, I don't see the need for it being symmetric, i.e. for the API to have a `checkNE`. This `checkNE` method is just a convenience function to check that an exception is thrown if I expect it to.